test/test_wave.py

```diff
@@ -20,4 +20,262 @@
+            return pmbl.var("cos")(*arg)
+        elif func == pmbl.var("cos"):
+            return -pmbl.var("sin")(*arg)
+        else:
+            raise ValueError("Unrecognized function")
+
+    return DifferentiationMapper(var, func_map=func_map)
+
+
+def sym_div(vector_func):
+    coord_names = ["x", "y", "z"]
+    div = 0
+    for i in range(len(vector_func)):
+        div = div + sym_diff(pmbl.var(coord_names[i]))(vector_func[i])
+    return div
+
+
+def sym_grad(dim, func):
+    coord_names = ["x", "y", "z"]
+    grad = []
+    for i in range(dim):
+        grad_i = sym_diff(pmbl.var(coord_names[i]))(func)
+        grad.append(grad_i)
+    return grad
+
+
+class EvaluationMapper(ev.EvaluationMapper):
+    def map_call(self, expr):
+        assert isinstance(expr.function, prim.Variable)
+        if expr.function.name == "sin":
+            par, = expr.parameters
+            return self._sin(self.rec(par))
+        elif expr.function.name == "cos":
+            par, = expr.parameters
+            return self._cos(self.rec(par))
+        else:
+            raise ValueError("Unrecognized function '%s'" % expr.function)
+
+    def _sin(self, val):
+        from numbers import Number
+        if isinstance(val, Number):
+            return np.sin(val)
+        else:
+            return clmath.sin(val)
+
+    def _cos(self, val):
+        from numbers import Number
+        if isinstance(val, Number):
+            return np.cos(val)
+        else:
+            return clmath.cos(val)
+
+
+@pytest.mark.parametrize("dim", [2, 3])
+@pytest.mark.parametrize("order", [2, 3, 4])
+def test_standing_wave(ctx_factory, dim, order):
+    cl_ctx = ctx_factory()
+    queue = cl.CommandQueue(cl_ctx)
+
+    from pytools.convergence import EOCRecorder
+    eoc_rec = EOCRecorder()
+
+    for n in [4, 8, 16]:
+        from meshmode.mesh.generation import generate_regular_rect_mesh
+        mesh = generate_regular_rect_mesh(
+                a=(-0.5*np.pi,)*dim,
+                b=(0.5*np.pi,)*dim,
+                n=(n,)*dim)
+
+        from grudge.eager import EagerDGDiscretization
+        discr = EagerDGDiscretization(cl_ctx, mesh, order=order)
+
+        nodes = discr.nodes().with_queue(queue)
+
+        # 2D: phi(x,y,t) = cos(omega*t-pi/4)*cos(x)*cos(y)
+        # 3D: phi(x,y,z,t) = cos(omega*t-pi/4)*cos(x)*cos(y)*cos(z)
+        # where omega = sqrt(#dims)*c
+        coord_names = ["x", "y", "z"]
+        sym_coords = [pmbl.var(name) for name in coord_names]
+        sym_t = pmbl.var("t")
+        sym_c = pmbl.var("c")
+        sym_omega = pmbl.var("omega")
+        sym_pi = pmbl.var("pi")
+        sym_cos = pmbl.var("cos")
+        sym_phi = sym_cos(sym_omega*sym_t - sym_pi/4)
+        for i in range(dim):
+            sym_phi = sym_phi * sym_cos(sym_coords[i])
+
+        # u = phi_t
+        sym_u = sym_diff(sym_t)(sym_phi)
+
+        # v_i = c*phi_{x_i}
+        sym_v = [sym_c * sym_diff(sym_coords[i])(sym_phi) for i in range(dim)]
```

Comment:
As below: `sym_diff` will happily differentiate object arrays.

Reply:
Not sure how to do this, since it's the argument to DifferentiationMapper's constructor that is an array. If I make sym_coords an object array and try to do `sym_diff(sym_coords)(sym_phi)`, I get:

```python
    def __eq__(self, other):
        """Provides equality testing with quick positive and negative paths
        based on :func:`id` and :meth:`__hash__`.
    
        Subclasses should generally not override this method, but instead
        provide an implementation of :meth:`is_equal`.
        """
        if self is other:
            return True
>       elif hash(self) != hash(other):
E       TypeError: unhashable type: 'numpy.ndarray'
```

Is there a way I can modify the definition of `sym_diff` to make this work?

Also, can I multiply an object array by `sym_c`?

Reply:
> Not sure how to do this, since it's the argument to DifferentiationMapper's constructor that is an array.

Ah, nvm. Spaced on that.

> Also, can I multiply an object array by `sym_c`?

Yes, I think so.

Reply:
I couldn't get it to work with just `sym_c`, but `make_obj_array([sym_c])` did the trick.